include/sys/notify.h

```diff
@@ -0,0 +1,361 @@
+ * function behavior and store completion information.
+ *
+ * @param result pointer to storage for the result of the operation.
+ * The result is stored only if the operation has completed.
+ *
+ * @retval 0 if the operation has completed.
+ * @retval -EAGAIN if the operation has not completed.
+ */
+static inline int sys_notify_fetch_result(const struct sys_notify *notify,
+					    int *result)
+{
+	__ASSERT_NO_MSG(notify != NULL);
+	__ASSERT_NO_MSG(result != NULL);
+	int rv = -EAGAIN;
+
+	if (sys_notify_get_method(notify) == SYS_NOTIFY_METHOD_COMPLETED) {
+		rv = 0;
+		*result = notify->result;
+	}
+	return rv;
+}
+
+/**
+ * @brief Initialize a notify object for spin-wait notification.
+ *
+ * Clients that use this initialization receive no asynchronous
+ * notification, and instead must periodically check for completion
+ * using sys_notify_fetch_result().
+ *
+ * On completion of the operation the client object must be
+ * reinitialized before it can be re-used.
+ *
+ * @param notify pointer to the notification configuration object.
+ */
+static inline void sys_notify_init_spinwait(struct sys_notify *notify)
+{
+	__ASSERT_NO_MSG(notify != NULL);
+
+	*notify = (struct sys_notify){
+		.flags = SYS_NOTIFY_METHOD_SPINWAIT,
+	};
+}
+
+/**
+ * @brief Initialize a notify object for (k_poll) signal notification.
+ *
+ * Clients that use this initialization will be notified of the
+ * completion of operations through the provided signal.
+ *
+ * On completion of the operation the client object must be
+ * reinitialized before it can be re-used.
+ *
+ * @note
+ *   @rst
+ *   This capability is available only when :option:`CONFIG_POLL` is
+ *   selected.
+ *   @endrst
+ *
+ * @param notify pointer to the notification configuration object.
+ *
+ * @param sigp pointer to the signal to use for notification.  The
+ * value must not be null.  The signal must be reset before the client
+ * object is passed to the on-off service API.
+ */
+static inline void sys_notify_init_signal(struct sys_notify *notify,
+					    struct k_poll_signal *sigp)
+{
+	__ASSERT_NO_MSG(notify != NULL);
+	__ASSERT_NO_MSG(sigp != NULL);
+
+	*notify = (struct sys_notify){
+		.method = {
+			.signal = sigp,
+		},
+		.flags = SYS_NOTIFY_METHOD_SIGNAL,
+	};
+}
+
+/**
+ * @brief Initialize a notify object for callback notification.
+ *
+ * Clients that use this initialization will be notified of the
+ * completion of operations through the provided callback.  Note that
+ * callbacks may be invoked from various contexts depending on the
+ * specific service; see @ref sys_notify_generic_callback.
+ *
+ * On completion of the operation the client object must be
+ * reinitialized before it can be re-used.
+ *
+ * @param notify pointer to the notification configuration object.
+ *
+ * @param handler a function pointer to use for notification.
+ */
+static inline void sys_notify_init_callback(struct sys_notify *notify,
+					      sys_notify_generic_callback handler)
```

Comment:
This looks like too long line, in this case I would just make sure the line is max 80 chars long even if the indentation would not match. Other option is just to put the 1st parameter to new line.

Reply:
Generally we don't do that sort of thing just to satisfy the line length test (which is advisory in Zephyr).